src/vcpkg/base/downloads.cpp

```diff
@@ -797,96 +316,114 @@ namespace vcpkg
+        CURL* curl = curl_easy_init();
+        if (!curl) Checks::unreachable(VCPKG_LINE_INFO);
 
-            return false;
+        curl_slist* request_headers = nullptr;
+        if (!raw_url.starts_with("ftp://"))
+        {
+            for (auto&& header : headers)
+                request_headers = curl_slist_append(request_headers, header.c_str());
         }
 
-        auto http_cmd = Command{"curl"}.string_arg("-X").string_arg(method);
-        add_curl_headers(http_cmd, headers);
-        http_cmd.string_arg("-w").string_arg("\\n" + guid_marker.to_string() + "%{http_code}");
-        http_cmd.string_arg(raw_url);
-        http_cmd.string_arg("-T").string_arg(file);
-        int code = 0;
-        auto res = cmd_execute_and_stream_lines(context, http_cmd, [&code](StringView line) {
-            if (line.starts_with(guid_marker))
-            {
-                code = std::strtol(line.data() + guid_marker.size(), nullptr, 10);
-            }
-        });
+        auto upload_url = url_encode_spaces(raw_url);
+        curl_easy_setopt(curl, CURLOPT_USERAGENT, vcpkg_curl_user_agent.data());
+        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, request_headers);
+        curl_easy_setopt(curl, CURLOPT_URL, upload_url.c_str());
+        curl_easy_setopt(curl, CURLOPT_UPLOAD, 1L);
+        curl_easy_setopt(curl, CURLOPT_READDATA, fileptr.get());
+        curl_easy_setopt(curl, CURLOPT_READFUNCTION, &read_file_callback);
+        curl_easy_setopt(curl, CURLOPT_INFILESIZE_LARGE, static_cast<curl_off_t>(file_size));
 
-        auto pres = res.get();
-        if (!pres)
+        auto result = curl_easy_perform(curl);
+        if (result != CURLE_OK)
         {
+            context.report_error(msg::format(msgCurlFailedGeneric,
+                                             msg::exit_code = static_cast<int>(result),
+                                             msg::error_msg = curl_easy_strerror(result)));
+            curl_easy_cleanup(curl);
+            curl_slist_free_all(request_headers);
             return false;
         }
 
-        if (*pres != 0 || (code >= 100 && code < 200) || code >= 300)
+        long response_code = 0;
+        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
+
+        if ((response_code >= 100 && response_code < 200) || response_code >= 300)
         {
-            context.report_error(msg::format(
-                msgCurlFailedToPutHttp, msg::exit_code = *pres, msg::url = sanitized_url, msg::value = code));
+            context.report_error(msg::format(msgCurlFailedToPutHttp,
+                                             msg::exit_code = static_cast<int>(result),
+                                             msg::error_msg = curl_easy_strerror(result),
+                                             msg::url = sanitized_url,
+                                             msg::value = response_code));
+            curl_easy_cleanup(curl);
+            curl_slist_free_all(request_headers);
```

Comment:
Ditto this suggests there should be an RAII type here.

Reply:
Also the loop to add headers is in here 3 times, add as a function on that RAII type?